x-pack/elastic-agent/pkg/core/monitoring/server/process.go

```diff
@@ -0,0 +1,180 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/elastic/beats/v7/metricbeat/mb/parse"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/artifact"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/monitoring/beats"
+)
+
+const (
+	processIDKey      = "processID"
+	monitoringSuffix  = "-monitoring"
+	separator         = "-"
+	timeout           = 10 * time.Second
+	errTypeUnexpected = "UNEXPECTED"
+
+	httpPlusPrefix = "http+"
+)
+
+var (
+	// ErrProgramNotSupported returned when requesting metrics for not supported program.
+	ErrProgramNotSupported = errors.New("specified program is not supported")
+)
+
+func processHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json; charset=utf-8")
+
+		vars := mux.Vars(r)
+		id, found := vars[processIDKey]
+
+		if !found {
+			writeResponse(
+				w,
+				unexpectedErrorWithReason("productID not found"),
+			)
+			return
+		}
+
+		metricsBytes, statusCode, metricsErr := processMetrics(r.Context(), id)
+		switch metricsErr {
+		case ErrProgramNotSupported:
+			w.WriteHeader(http.StatusNotFound)
+		default:
+			w.WriteHeader(statusCode)
+		}
+
+		if metricsErr != nil {
+			writeResponse(w, unexpectedErrorWithReason("failed fetching metrics: %s", metricsErr.Error()))
+			return
+		}
+
+		fmt.Fprint(w, string(metricsBytes))
+	}
+}
+
+func processMetrics(ctx context.Context, id string) ([]byte, int, error) {
```

Comment:
I like the idea of combining the status code and the error a real error type. HTTP errors would report the code and message which then gets written to the response by the top level handle. We would use InternalError by default if it is 'any' error not implementing our StatusError interface. For example see this blog post: https://blog.questionable.services/article/http-handler-error-handling-revisited/#a-little-different